src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -1183,6 +1191,26 @@ protected override BoundNode VisitReturnStatementNoAdjust(BoundReturnStatement n
             return null;
         }
 
+        private TypeWithState VisitRefExpression(BoundExpression expr, TypeSymbolWithAnnotations destinationType)
+        {
+            Visit(expr);
+            TypeWithState resultType = ResultType;
+            if (!expr.IsSuppressed && RemoveConversion(expr, includeExplicitConversions: false).expression.Kind != BoundKind.ThrowExpression)
+            {
+                var lvalueResultType = LvalueResultType;
```

Comment:
I don't really follow this code. When could LValueResultType be different from ResultType?

Reply:
You can have a non-nullable ref local `ref string x` which has a null-state of MaybeNull because it was initialized or assigned with an expression that had a null-state of MaybeNull.

---
In reply to: [262688367](https://github.com/dotnet/roslyn/pull/33869#discussion_r262688367) [](ancestors = 262688367)

Reply:
Here's the scenario that hits that (warning on //2):
```C#
[Fact, WorkItem(31297, "[https://github.com/dotnet/roslyn/issues/31297](https://github.com/dotnet/roslyn/issues/31297)")]
        public void SuppressNullableWarning_RefReassignment()
        {
            var comp = CreateCompilation(@"
class C
{
    void M()
    {
        ref string x = ref NullableRef(); // 1
        ref string x2 = ref x; // 2
        ref string x3 = ref x!;

        ref string y = ref NullableRef()!;
        ref string y2 = ref y;
        ref string y3 = ref y!;
    }

    ref string? NullableRef() => throw null!;
}", options: WithNonNullTypesTrue());

            comp.VerifyDiagnostics(
                // (6,28): warning CS8619: Nullability of reference types in value of type 'string?' doesn't match target type 'string'.
                //         ref string x = ref NullableRef(); // 1
                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "NullableRef()").WithArguments("string?", "string").WithLocation(6, 28),
                // (7,29): warning CS8597: Possible null value.
                //         ref string x2 = ref x; // 2
                Diagnostic(ErrorCode.WRN_PossibleNull, "x").WithLocation(7, 29)
                );
        }
```

---
In reply to: [262691050](https://github.com/dotnet/roslyn/pull/33869#discussion_r262691050) [](ancestors = 262691050,262688367)